token/program-2022/src/extension/reallocate.rs

```diff
@@ -72,19 +74,42 @@ pub fn process_reallocate(
     // if additional lamports needed to remain rent-exempt, transfer them
     let rent = Rent::get()?;
     let new_minimum_balance = rent.minimum_balance(needed_account_len);
-    let lamports_diff = new_minimum_balance.saturating_sub(token_account_info.lamports());
-    invoke(
-        &system_instruction::transfer(payer_info.key, token_account_info.key, lamports_diff),
-        &[
-            payer_info.clone(),
-            token_account_info.clone(),
-            system_program_info.clone(),
-        ],
-    )?;
+
+    let current_lamport_reserve = token_account_info
+        .lamports()
+        .checked_sub(native_amount.unwrap_or(0))
+        .ok_or(TokenError::Overflow)?;
+    let lamports_diff = new_minimum_balance.saturating_sub(current_lamport_reserve);
+    if lamports_diff > 0 {
+        invoke(
+            &system_instruction::transfer(payer_info.key, token_account_info.key, lamports_diff),
+            &[
+                payer_info.clone(),
+                token_account_info.clone(),
+                system_program_info.clone(),
+            ],
+        )?;
+    }
 
     // unpack to set account_type, if needed
-    let mut token_account = token_account_info.data.borrow_mut();
-    set_account_type::<Account>(&mut token_account)?;
+    let mut token_account_data = token_account_info.data.borrow_mut();
+    set_account_type::<Account>(&mut token_account_data)?;
+
+    // sync the rent exempt reserve for native accounts
+    let mut token_account = StateWithExtensionsMut::<Account>::unpack(&mut token_account_data)?;
+    if token_account.base.is_native() {
+        // sanity check that there are enough lamports to cover the token amount
+        // and the rent exempt reserve
+        let native_amount = native_amount.ok_or(TokenError::InvalidState)?;
```

Comment:
These 2 lines are technically redundant, since we've already checked `is_native()` to construct `native_amount`.
Could do:
```suggestion
    if let Some(native_amount) = native_amount {
        // sanity check that there are enough lamports to cover the token amount
        // and the rent exempt reserve
```
I feel ambivalent about it.

Reply:
Ah that is neater, thanks! I just didn't do an unwrap to avoid a conversation with the auditor